File: task4/matrixSlicing.py
import numpy as np
import matplotlib.pyplot as plt

rows = int(input("Input number of rows in the matrix: "))
cols = int(input("Input number of columns in the matrix: "))

matrix = np.random.randint(0, 11, size=(rows, cols))
print("\nGenerated Matrix:\n", matrix)

print("Enter the slice ranges for rows and columns (start inclusive, end exclusive):")
row_start = int(input(f"Start row (0 to {rows-1}): "))
row_end = int(input(f"End row (1 to {rows}): "))
col_start = int(input(f"Start column (0 to {cols-1}): "))
col_end = int(input(f"End column (1 to {cols}): "))

# Swap if end < start
if row_end < row_start:
    row_start, row_end = row_end, row_start
if col_end < col_start:
    col_start, col_end = col_end, col_start

sliced_matrix = matrix[row_start:row_end, col_start:col_end]
print("\nSliced Submatrix:\n", sliced_matrix)

#original matrix
plt.figure(figsize=(10, 4))
plt.subplot(1, 2, 1)
plt.title("Original Matrix")
plt.imshow(matrix, cmap="viridis", interpolation="nearest")
plt.colorbar()

#sliced matrix
plt.subplot(1, 2, 2)
plt.title("Sliced Submatrix")
plt.imshow(sliced_matrix, cmap="viridis", interpolation="nearest")
plt.colorbar()

plt.show()

File: task4/matrixSlicing.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main() {

    int rows, cols;
    printf("Input number of rows in the matrix: ");
    scanf("%d", &rows);
    printf("Input number of columns in the matrix: ");
    scanf("%d", &cols);

    int **matrix = malloc(rows * sizeof(int*));
    for (int i = 0; i < rows; i++)
        matrix[i] = malloc(cols * sizeof(int));

    for (int i = 0; i < rows; i++)
        for (int j = 0; j < cols; j++)
            matrix[i][j] = rand() % 11;

    printf("\nGenerated Matrix:\n");
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++)
            printf("%4d", matrix[i][j]);
        printf("\n");
    }

    int row_start, row_end, col_start, col_end;
    printf("Enter the slice ranges for rows and columns (start inclusive, end exclusive):\n");
    printf("Start row (0 to %d): ", rows-1);
    scanf("%d", &row_start);
    printf("End row (1 to %d): ", rows);
    scanf("%d", &row_end);
    printf("Start column (0 to %d): ", cols-1);
    scanf("%d", &col_start);
    printf("End column (1 to %d): ", cols);
    scanf("%d", &col_end);

    // Swap if end < start
    if (row_end < row_start) {
        int temp = row_start;
        row_start = row_end;
        row_end = temp;
    }
    if (col_end < col_start) {
        int temp = col_start;
        col_start = col_end;
        col_end = temp;
    }

    printf("\nSliced Submatrix:\n");
    for (int i = row_start; i < row_end; i++) {
        for (int j = col_start; j < col_end; j++)
            printf("%4d", matrix[i][j]);
        printf("\n");
    }

    for (int i = 0; i < rows; i++)
        free(matrix[i]);
    free(matrix);

    return 0;
}
